source/simulator/assembly.cc

```diff
@@ -1701,6 +1807,839 @@ namespace aspect
+        );
+
+    const FEValuesExtractors::Scalar solution_field
+      = (advection_field.is_temperature()
+         ?
+         introspection.extractors.temperature
+         :
+         introspection.extractors.compositional_fields[advection_field.compositional_variable]
+        );
+
+    //loop over all possible subfaces of the cell, and reset their matrices.
+    for (unsigned int f = 0; f < GeometryInfo<dim>::max_children_per_face * GeometryInfo<dim>::faces_per_cell; ++f)
+      {
+        data.local_matrices_ext_int[f] = 0;
+        data.local_matrices_int_ext[f] = 0;
+        data.local_matrices_ext_ext[f] = 0;
+      }
+
+    for (unsigned int f = 0; f < GeometryInfo<dim>::faces_per_cell; ++f)
+      {
+
+        typename DoFHandler<dim>::face_iterator face = cell->face (f);
+        if (face->at_boundary())
+          {
+            scratch.face_finite_element_values.reinit (cell, f);
+
+            scratch.face_finite_element_values[introspection.extractors.velocities].get_function_values(current_linearization_point,
+                scratch.face_current_velocity_values);
+
+            //get the mesh velocity, as we need to subtract it off of the advection systems
+            if (parameters.free_surface_enabled)
+              scratch.face_finite_element_values[introspection.extractors.velocities].get_function_values(free_surface->mesh_velocity,
+                  scratch.face_mesh_velocity_values);
+
+            compute_material_model_input_values (current_linearization_point,
+                                                 scratch.face_finite_element_values,
+                                                 cell,
+                                                 true,
+                                                 scratch.face_material_model_inputs);
+            material_model->evaluate(scratch.face_material_model_inputs,
+                                     scratch.face_material_model_outputs);
+
+            HeatingModel::HeatingModelOutputs face_heating_model_outputs(n_q_points, parameters.n_compositional_fields);
+            heating_model_manager.evaluate(scratch.face_material_model_inputs,
+                                           scratch.face_material_model_outputs,
+                                           face_heating_model_outputs);
+
+            if (((parameters.fixed_temperature_boundary_indicators.find(cell->face(f)->boundary_id()) != parameters.fixed_temperature_boundary_indicators.end()) && (advection_field.is_temperature()))
+                ||
+                (!advection_field.is_temperature()))
+              {
+                //We are either in the case of a Dirichlet temperature boundary, or we are assembling for a composition variable. In the former, impose the Dirichlet value weakly using a matrix term and RHS term. In the latter, if this is an inflow boundary then impose a zero Dirichlet value by assembling the corresponding matrix term but no RHS term
```

Comment:
In the composition case, Dirichlet boundary conditions are only imposed on the inflow boundary, because it is a pure advection equation. I've implemented this following the equations stated in the manual (§2.1.2), which say that the Dirichlet conditions are homogeneous. I’m aware that there are actually fixed_composition_boundary_indicators implemented, so perhaps I should change this to handle these as well. If so, I think the manual might want to be updated to reflect this. However it would probably be wise to explicitly state that Dirichlet conditions can only be imposed on the inflow boundaries, regardless of the boundary indicators passed. What do you think - should I implement this?